### PR TITLE
made hostanalyzermanager to use object as id type

### PR DESCRIPTION
**Pull request by heejaechang**
changed AnalyzerReference.Id to object type.

and made corresponding change to HostAnalyzerManager to store it as object.

also, made IDiagnosticService to return <displayname, analyzer reference> map as it used to

### CODE REVIEW

**Comment by heejaechang** [comment]
@tmeschter @srivatsn @shyamnamboodiripad can you take a look, please?

**Comment by shyamnamboodiripad** [comment]
Please ping the compat folks since this is a public API change...

File: src/Compilers/Core/Portable/PublicAPI.txt [hunk 1]
```diff
@@ -1883,6 +1987,9 @@ override sealed Microsoft.CodeAnalysis.Diagnostics.DiagnosticAnalyzer.ToString()
 override sealed Microsoft.CodeAnalysis.LocalizableString.Equals(object other) -> bool
 override sealed Microsoft.CodeAnalysis.LocalizableString.GetHashCode() -> int
 override sealed Microsoft.CodeAnalysis.LocalizableString.ToString() -> string
+readonly Microsoft.CodeAnalysis.Diagnostics.AnalyzerLoadFailureEventArgs.ErrorCode -> Microsoft.CodeAnalysis.Diagnostics.AnalyzerLoadFailureEventArgs.FailureErrorCode
+readonly Microsoft.CodeAnalysis.Diagnostics.AnalyzerLoadFailureEventArgs.Exception -> System.Exception
+readonly Microsoft.CodeAnalysis.Diagnostics.AnalyzerLoadFailureEventArgs.TypeName -> string
```
Comment (shyamnamboodiripad) [inline src/Compilers/Core/Portable/PublicAPI.txt]: Are these just lines that moved? Looks like we have bug in the public API fixer (shouldn't it ensure that this file is always sorted)? cc @tmeschter

Comment (heejaechang) [inline src/Compilers/Core/Portable/PublicAPI.txt]: @shyamnamboodiripad do you know alias in github for compat folks?

by the way, did we even got approval for the "adding Id" at the first place?

**Comment by heejaechang** [comment]
@dotnet-bot test please

**Comment by heejaechang** [comment]
@dotnet-bot test this please

**Comment by heejaechang** [comment]
@dotnet-bot retest this please

**Comment by heejaechang** [comment]
@dotnet-bot retest please

**Comment by tmeschter** [comment]
@dotnet-bot Retest this, please.

**Comment by heejaechang** [comment]
@tmeschter @srivatsn @shyamnamboodiripad can you take a look, please?

**Comment by heejaechang** [comment]
I am checking this in to unblock @tmeschter 

I will address PR feedback in follow up PR.

File: src/Features/Core/Diagnostics/DiagnosticAnalyzerService.cs [hunk 2]
```diff
@@ -53,10 +53,53 @@ protected DiagnosticAnalyzerService(ImmutableArray<AnalyzerReference> workspaceA
...
+            return builder.ToImmutable();
+        }
+
+        private static string GetDisplayName(Dictionary<string, int> nameCollisionMap, AnalyzerReference reference)
+        {
+            var name = reference.Display ?? FeaturesResources.Unknown;
+            var count = GetNameCollisionCount(nameCollisionMap, name);
+            return count == 0 ? name : $"{name} [#{count}]";
+        }
+
```
Comment (tmeschter) [inline src/Features/Core/Diagnostics/DiagnosticAnalyzerService.cs]: @heejaechang This naming is going to cause other problems with the rule set editor. Instead, if multiple `AnalyzerReferences` have the same `DisplayName` just union together their diagnostic descriptors.

**Comment by tmeschter** [comment]
@heejaechang LGTM. Thanks!

**Pull request merged**